### PR TITLE
Explicit registration of canvas-specific tool subclasses.

### LINKED ISSUE #17986: MEP22 per-backend tool registration
I'll try to split off #8712 into smaller parts to perhaps keep the discussion more focused.

Currently, tools that have backend-specific implementations are "registered" by defining them in their own backend module, and then overwriting the "original" class in the `backend_tools`.  For example, `backend_qt5.py` contains
```python
class SaveFigureQt(backend_tools.SaveFigureBase): ...
class ConfigureSubplotsQt(backend_tools.ConfigureSubplotsBase): ...
...

backend_tools.ToolSaveFigure = SaveFigureQt
backend_tools.ToolConfigureSubplots = ConfigureSubplotsQt
...
```
and likewise in `backend_gtk3`, etc.  Unfortunately, this means that things won't work if multiple backend modules get imported in the same process (unless they are imported in the right order).  This is not a theoretical concern: multiple backend modules do indeed get imported in the same process when performing automatic backend fallback detection.  (It is true that the actually used backend is *typically* the last one imported but this is certainly not something we guarantee in the future.)  Therefore, I believe there should instead be some other registration approach that links the tool classes to the backends (such as the one proposed in #9941, but I'm not particularly wedded to this approach either).  (Note that one cannot rely either on heuristics such as "the Tool defined in the same module as the backend", because e.g. the qt5 tools are defined in backend_qt5.py but the backend modules are in backend_qt5agg.py/backend_qt5cairo.py or even third-party qt5 backends.)

Extra points if the tool registration mechanism allows *any* tool class to be overridden on a per-backend basis.

### PR DESCRIPTION
**Pull request by anntzer**
Don't mess up with overwriting backend_tools globals, which would fail if multiple backend modules are imported.

This proposes an API to fix #17986 (and therefore unblock https://github.com/matplotlib/matplotlib/pull/20273#issuecomment-845703009).  I had tried another approach back in #9941, but that PR did many more cleanups at the same time and therefore got bogged down in side discussions; this one solely implements canvas-specific tool subclass registration.  That registration mechanism is private for now, but we can always consider making it public if there's demand for it.  (I actually guess there would not be that much demand for third-party tool implementers because even if the tool really required some backend-specific logic, it could just use a plain old if-switch instead (https://github.com/anntzer/mplinorm/blob/dbd0b34ab4b2ad3799410565ec6ad31f992b7576/lib/mplinorm.py#L162) because a single module/package would provide the tools; the dispatch problem on Matplotlib's side mostly comes from the fact that each backend implements its own tools in its own module.  The only case where this would really be necessary for third parties is for integration into a new event loop.)

(See the linked PRs and issues for more details about the problem this is solving.)

attn @fariza, I guess.

## PR Checklist

- [ ] Has pytest style unit tests (and `pytest` passes).
- [ ] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (run `flake8` on changed files to check).
- [ ] New features are documented, with examples if plot related.
- [ ] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).
- [ ] Conforms to Matplotlib style conventions (install `flake8-docstrings` and run `flake8 --docstring-convention=all`).
- [ ] New features have an entry in `doc/users/next_whats_new/` (follow instructions in README.rst there).
- [ ] API changes documented in `doc/api/next_api_changes/` (follow instructions in README.rst there).

### CODE REVIEW

**Comment by fariza** [comment]
Does the example still works?
https://github.com/matplotlib/matplotlib/blob/master/examples/user_interfaces/toolmanager_sgskip.py

**Comment by anntzer** [comment]
Yes, it works.

File: lib/matplotlib/backend_tools.py [hunk 1]
```diff
@@ -36,6 +37,35 @@ class Cursors(enum.IntEnum):  # Must subclass int for the macOS backend.
...
+# of the requested tool class has been registered (either for the current
+# canvas class or for a parent class), in which case that tool subclass will be
+# instantiated instead.  This is the mechanism used e.g. to allow different
+# GUI backends to implement different specializations for ConfigureSubplots.
+
+
+_tool_registry = set()
+
+
+def _register_tool_class(canvas_cls, tool_cls=None):
```
Comment (fariza) [inline lib/matplotlib/backend_tools.py:52]: I don't see how to do it easily, but I still need to ask, is it possible to add test to these two functions?

Comment (anntzer) [reply lib/matplotlib/backend_tools.py:52]: They are covered already by existing tests (https://codecov.io/gh/matplotlib/matplotlib/compare/971d1106464560038151bf79d032515cd843b149...49d855f552bb44282b40ff689c5de5fe60ffbc15/diff).

**Review by fariza — APPROVED** [review]

**Comment by fariza** [comment]
@anntzer does it need a rebase?

**Comment by anntzer** [comment]
rebased

**Comment by anntzer** [comment]
Please note that the standard policy is *two* positive reviews.

**Comment by fariza** [comment]
Sorry, I saw Github requesting only one review, and I saw this as kind of small well contained change

**Pull request merged**